### PR TITLE
#11 create endpoint to get buildings to build/upgrade with requirements

### LINKED ISSUE #11: Create endpoint to get buildings to build/upgrade with requirements

### PR DESCRIPTION
**Pull request by Skrabka98**

### CODE REVIEW

File: TribalWars/src/ApplicationCore/Entities/Building.cs [hunk 1]
```diff
@@ -19,5 +19,6 @@ public class Building
         public virtual ICollection<BuildingRequiredBuilding> BuildingRequiredBuildings { get; set; }
         public virtual ICollection<VillageBuilding> VillageBuildings { get; set; }
         public virtual ICollection<BuildingRequiredMaterial> BuildingRequiredMaterials { get; set; }
+        public virtual RequiredBuilding RequiredBuildings { get; set; }
```
Comment (andrzejwojciak) [inline TribalWars/src/ApplicationCore/Entities/Building.cs]: Field isn't a collection but name is in plural form.

File: TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs [hunk 2]
```diff
@@ -10,8 +10,8 @@ public class BuildingRequiredBuilding
     {
         public int Id { get; set; }
         public int BuildingId { get; set; }
-        public int RequiredBuildingId { get; set; }
-        public int Level { get; set; }
         public virtual Building Building { get; set; }
+        public int RequiredBuildingId { get; set; }
+        public virtual RequiredBuilding RequiredBuilding { get; set; }
```
Comment (andrzejwojciak) [inline TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs:15]: It will be more readable if you will separate entity values from other entities inclusions, e.q.:
`public int Id { get; set; }`
`public int BuildingId { get; set; }`
`public int RequiredBuildingId { get; set; }`

`public virtual RequiredBuilding RequiredBuilding { get; set; }`
`public virtual Building Building { get; set; }`

File: TribalWars/src/ApplicationCore/Entities/RequiredBuilding.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace ApplicationCore.Entities
+{
+    public class RequiredBuilding
+    {
+        public int Id { get; set; }
+        public int ReqBuildingId { get; set; }
+        public int Level { get; set; }
+        public virtual ICollection<BuildingRequiredBuilding> Buildings { get; set; }
+        public virtual Building ReqBuilding { get; set; }
```
Comment (andrzejwojciak) [inline TribalWars/src/ApplicationCore/Entities/RequiredBuilding.cs]: Unnecessary shortcut in ReqBuilding and name of collection Buildings does not apply to Building entity.

File: TribalWars/src/Infrastructure/Configuration/BuildingRequiredBuildingConfig.cs [hunk 4]
```diff
@@ -18,6 +13,19 @@ public void Configure(EntityTypeBuilder<BuildingRequiredBuilding> builder)
...
+                .HasForeignKey(buildingRequiredBuilding => buildingRequiredBuilding.RequiredBuildingId);
+
+            builder.HasOne(rbr => rbr.RequiredBuilding)
+                    .WithMany(rb => rb.Buildings)
+                    .OnDelete(DeleteBehavior.Restrict);
+
+            builder.HasOne(rbr => rbr.Building)
+                    .WithMany(rb => rb.BuildingRequiredBuildings)
+                    .OnDelete(DeleteBehavior.Cascade);
+
```
Comment (andrzejwojciak) [inline TribalWars/src/Infrastructure/Configuration/BuildingRequiredBuildingConfig.cs]: Declarations about relationships are duplicated.

File: TribalWars/src/Infrastructure/Configuration/BuildingRequiredConfig.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infrastructure.Configuration
+{
+    class BuildingRequiredConfig : IEntityTypeConfiguration<RequiredBuilding>
```
Comment (andrzejwojciak) [inline TribalWars/src/Infrastructure/Configuration/BuildingRequiredConfig.cs]: Entity has name "RequiredBuilding" and configuration class has name "BuildingRequired", this is not consistent.

File: TribalWars/src/Infrastructure/Mapping/ModelToResourceProfile.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+using Infrastructure.ViewModels;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infrastructure.Mapping
+{
+    public class ModelToResourceProfile : Profile
```
Comment (andrzejwojciak) [inline TribalWars/src/Infrastructure/Mapping/ModelToResourceProfile.cs]: Class name is suggesting we can find here Model to Resource mapping profile, but in fact it's a Entity to ViewModel profile.

File: TribalWars/src/Infrastructure/ViewModels/BuildingRequiredBuildingViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+﻿using ApplicationCore.Entities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infrastructure.ViewModels
```
Comment (andrzejwojciak) [inline TribalWars/src/Infrastructure/ViewModels/BuildingRequiredBuildingViewModel.cs]: Classes like ViewModels belong to domain project, so should be located in ApplicationCore.

File: TribalWars/src/WebApi/Controllers/RequiredBuildingsController.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace WebApi.Controllers
+{
+    [ApiController]
+    public class RequiredBuildingsController : Controller
+    {
+        private readonly IBuildingRequiredService _buildingRequiredService;
+        private readonly IMapper _mapper;
```
Comment (andrzejwojciak) [inline TribalWars/src/WebApi/Controllers/RequiredBuildingsController.cs:16]: Missing separation between fields and constructor.

File: TribalWars/src/WebApi/WebApi.csproj [hunk 9]
```diff
@@ -26,4 +26,6 @@
       <Folder Include="Settings" />
     </ItemGroup>
 
+    <ProjectExtensions><VisualStudio><UserProperties appsettings_1json__JsonSchema="" /></VisualStudio></ProjectExtensions>
```
Comment (andrzejwojciak) [inline TribalWars/src/WebApi/WebApi.csproj]: Why is this added here?

File: TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs (hunk 2 above)
Comment (Skrabka98) [reply TribalWars/src/ApplicationCore/Entities/BuildingRequiredBuilding.cs:15]: As Szymon explained to me, he said that such a division is clearer because it is known what it is used for.

File: TribalWars/src/ApplicationCore/Entities/RequiredBuilding.cs (hunk 3 above)
Comment (Skrabka98) [reply TribalWars/src/ApplicationCore/Entities/RequiredBuilding.cs]: The shortcut is because I cannot name it RequiredBuilding

Comment (HerringTheCoder) [reply TribalWars/src/ApplicationCore/Entities/RequiredBuilding.cs]: Can we think of another name that describes this property intention? I'm not a fan of this shortcut too.

Comment (Skrabka98) [reply TribalWars/src/ApplicationCore/Entities/RequiredBuilding.cs]: Maybe just Building and BuildingId?
